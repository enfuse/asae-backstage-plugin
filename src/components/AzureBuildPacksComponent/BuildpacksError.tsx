import React from 'react'
import * as Constants from '../../constants'

import Alert from '@material-ui/lab/Alert';
import { AlertTitle } from '@material-ui/lab';

type Error = {
  name: string;
  message: string;
  stack?: string;
};

export const BuildpacksError = (props : {error :Error} ) => {
    return <Alert severity="error" variant='outlined'>
            <AlertTitle>Possible Misconfiguration, check these values in your app-config.yml</AlertTitle>
            {Constants.ASAE_SERVICE_NAME }  <br/>
            {Constants.ASAE_RESOURCE_GROUP }  <br/>
            {Constants.ASAE_BUILD_SERVICE_NAME } <br/>
            {Constants.CONFIG_ASAE_SUBSCRIPTION_ID }  <br/>
            {Constants.CONFIG_ASAE_CREDENTIALS_CLIENT_ID }  <br/>
            {Constants.CONFIG_ASAE_CREDENTIALS_TENTANT_ID }  <br/><br/>
            {props.error.message}
          </Alert>
}