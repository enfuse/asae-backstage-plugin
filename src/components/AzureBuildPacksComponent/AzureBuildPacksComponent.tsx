import React from 'react';
import useAsync from 'react-use/lib/useAsync';
import {Table,
        TableBody,
        TableCell,
        TableHead,
        TableRow,
        Paper} from '@mui/material';
import { useStyles } from '../common/styles';
import { withAzureClient,DataProps} from '../common/withAzureClient';
import {TitleWithFilterSearch} from '../common/TitleWithFilterSearch'
import { BuildpacksError } from '../AzureBuildPacksComponent/BuildpacksError';
import * as Utils from '../../utils/asae-utils'
import * as Types from '../../types'
import { TableContainer } from '@material-ui/core';


export const AzureBuildPacksComponent = ({value}: DataProps) => {
  const classes = useStyles()
  const [search, setSearch] = React.useState('')
  const handleSearch = (e) => { setSearch(e.target.value)}
  return (
    <>
    <TitleWithFilterSearch  handleSearch={handleSearch} title="Available Buildpacks" />
    <TableContainer component={Paper}>
      <Table aria-label='Buildpacks'>
        <TableHead>
          <TableRow>
            <TableCell className={classes.tableCell} >Buildpack</TableCell>
            <TableCell className={classes.tableCell} >Builder</TableCell>
            <TableCell className={classes.tableCell} >Supported Languages</TableCell>
          </TableRow>
        </TableHead>
        
        <TableBody>
          {value.map(item => {
            if(item?.buildpack?.toLowerCase().includes(search.toLowerCase()) ||
             item?.builder?.toLowerCase().includes(search.toLowerCase()) ||
             item?.languages?.toLowerCase().includes(search.toLowerCase()) )
            return <TableRow className={classes.tableRow} key={item.buildpack+item.builder}>
                <TableCell className={classes.tableCell}>{item.buildpack}</TableCell>
                <TableCell className={classes.tableCell}>{item.builder}</TableCell>
                <TableCell className={classes.tableCell}>{item.languages}</TableCell>
              </TableRow>
          })}
        </TableBody>
      </Table>
    </TableContainer>
          </>
  )
}
const azureBuildpacksCallback = (client : any, entityAsaeInfo: Types.EntityAsae) => {
  const { value, loading, error } = useAsync(async (): Promise<Types.SupportedBuildpacks[]> => {
      const builders: Types.AsaeBuilder[]= []
      const builderIterator = client.buildServiceBuilder.list(entityAsaeInfo.resourceGroup, 
        entityAsaeInfo.asaeService, 
        entityAsaeInfo.buildServiceName)
      for await (const val of builderIterator){
        builders.push(val)
      }
      const buildpacksBuilders = builders.map((builder:Types.AsaeBuilder)=>{
        const buildpacks:Types.AsaeBuildpack[][] = []
        builder.properties.buildpackGroups.map(group=>buildpacks.push(group.buildpacks.flat() as Types.AsaeBuildpack[]))
        return {
          name: builder.name,
          buildpacks: buildpacks.flat()
        }
      })
      const buildPacks : Types.SupportedBuildpacks[][] = buildpacksBuilders.map(({name, buildpacks}): Types.SupportedBuildpacks[] => {
       return buildpacks.map((buildpack : Types.AsaeBuildpack) : Types.SupportedBuildpacks => {
          return {
            builder: name,
            buildpack: buildpack.id,
            languages: Utils.getCompatibleLanguages(buildpack.id)
          }
        })
      })
        return buildPacks.flat()
    },[]);
    return { value, loading, error }
}
export const AsaeBuildPacksWithAzureClient = withAzureClient(AzureBuildPacksComponent, 
                                                          BuildpacksError,
                                                          azureBuildpacksCallback)
