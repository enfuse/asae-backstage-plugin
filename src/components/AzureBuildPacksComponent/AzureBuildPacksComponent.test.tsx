import React from 'react';
import { render, RenderResult, screen } from '@testing-library/react';
import { AzureBuildPacksComponent } from './AzureBuildPacksComponent';
// import { rest } from 'msw';

describe('When DataTable renders', () => {
  let rendered : HTMLElement | undefined = undefined 

  const setup = () => {

      const testData = [{
          buildpack: 'item.buildpack',
          builder: 'item.builder',
          languages: 'item.languages'
        },{
          buildpack: 'item.buildpack2',
          builder: 'item.builder2',
          languages: 'item.languages2'
        }]

      const {baseElement} =  render(<AzureBuildPacksComponent value={testData}/>)
      rendered = baseElement
  }
  beforeEach(()=>{
    setup()
  })
  it('should match previous snapshot', async() => {
     
      expect(rendered).toMatchSnapshot();
  })
  it('should display a heading', async ()=>{
      expect(screen.getByRole('heading')).toBeInTheDocument()
  })
  it('should display 3 columns', async() => {
      expect(screen.getAllByRole('columnheader').length).toBe(3)
  })
  it('display table with data', async()=>{    
      expect(screen.getByRole('cell', {name:"item.buildpacks"})).toBeInTheDocument()
      expect(screen.getByRole('cell', {name:"item.buildpack2"})).toBeInTheDocument()
      expect(screen.getByRole('cell', {name:"item.builder"})).toBeInTheDocument()
      expect(screen.getByRole('cell', {name:"item.builder2"})).toBeInTheDocument()
      expect(screen.getByRole('cell', {name:"item.languages"})).toBeInTheDocument()
      expect(screen.getByRole('cell', {name:"item.languages"})).toBeInTheDocument()
  })
})
