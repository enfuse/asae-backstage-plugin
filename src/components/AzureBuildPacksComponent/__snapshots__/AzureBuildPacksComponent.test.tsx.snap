// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When DataTable renders should match previous snapshot 1`] = `
<body>
  <div>
    <div
      class="BackstageTable-root-1"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation2 MuiPaper-rounded"
        style="position: relative; width: 100%;"
      >
        <div
          class="MuiToolbar-root MuiToolbar-regular ForwardRef(MTableToolbarRef)-root-11 BackstageTableToolbar-root-8 MuiToolbar-gutters"
        >
          <div
            class="ForwardRef(MTableToolbarRef)-title-15 BackstageTableToolbar-title-9"
          >
            <h2
              class="MuiTypography-root MuiTypography-h5"
            >
              Supported ASAE Buildpacks
            </h2>
          </div>
          <div
            class="ForwardRef(MTableToolbarRef)-spacer-13"
          />
          <div
            class="MuiFormControl-root MuiTextField-root ForwardRef(MTableToolbarRef)-searchField-16 BackstageTableToolbar-searchField-10"
          >
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedStart MuiInputBase-adornedEnd"
            >
              <div
                class="MuiInputAdornment-root MuiInputAdornment-positionStart"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                  focusable="false"
                  title="Search"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M10 18h4v-2h-4v2zM3 6v2h18V6H3zm3 7h12v-2H6v2z"
                  />
                </svg>
              </div>
              <input
                aria-invalid="false"
                aria-label="Search"
                class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedStart MuiInputBase-inputAdornedEnd"
                placeholder="Filter"
                type="text"
                value=""
              />
              <div
                class="MuiInputAdornment-root MuiInputAdornment-positionEnd"
              >
                <button
                  aria-label="Clear Search"
                  class="MuiButtonBase-root MuiIconButton-root Mui-disabled Mui-disabled"
                  disabled=""
                  tabindex="-1"
                  type="button"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <svg
                      aria-hidden="true"
                      aria-label="clear"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                      />
                    </svg>
                  </span>
                </button>
              </div>
            </div>
          </div>
          <div
            class="ForwardRef(MTableToolbarRef)-actions-14"
          >
            <div>
              <div
                style="display: flex;"
              >
                <span>
                  <div
                    style="display: flex;"
                  />
                </span>
              </div>
            </div>
          </div>
        </div>
        <div
          class="Component-horizontalScrollContainer-18"
          style="overflow-x: auto; position: relative;"
        >
          <div>
            <div
              style="overflow-y: auto;"
            >
              <div>
                <table
                  class="MuiTable-root"
                  style="table-layout: auto;"
                >
                  <thead
                    class="MuiTableHead-root"
                  >
                    <tr
                      class="MuiTableRow-root MuiTableRow-head"
                    >
                      <th
                        class="MuiTableCell-root MuiTableCell-head ForwardRef(MTableHeaderRef)-header-20 BackstageTableHeader-header-19 MuiTableCell-alignLeft"
                        scope="col"
                        style="text-transform: uppercase; box-sizing: border-box;"
                      >
                        <span
                          aria-disabled="false"
                          class="MuiButtonBase-root MuiTableSortLabel-root"
                          role="button"
                          tabindex="0"
                        >
                          <div
                            aria-describedby="rbd-hidden-text-0-hidden-text-0"
                            data-rbd-drag-handle-context-id="0"
                            data-rbd-drag-handle-draggable-id="0"
                            data-rbd-draggable-context-id="0"
                            data-rbd-draggable-id="0"
                            draggable="false"
                            role="button"
                            tabindex="0"
                          >
                            Buildpack
                          </div>
                          <svg
                            aria-hidden="true"
                            class="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
                            focusable="false"
                            viewBox="0 0 24 24"
                          >
                            <path
                              d="M4 12l1.41 1.41L11 7.83V20h2V7.83l5.58 5.59L20 12l-8-8-8 8z"
                            />
                          </svg>
                        </span>
                      </th>
                      <th
                        class="MuiTableCell-root MuiTableCell-head ForwardRef(MTableHeaderRef)-header-20 BackstageTableHeader-header-19 MuiTableCell-alignLeft"
                        scope="col"
                        style="text-transform: uppercase; box-sizing: border-box;"
                      >
                        <span
                          aria-disabled="false"
                          class="MuiButtonBase-root MuiTableSortLabel-root"
                          role="button"
                          tabindex="0"
                        >
                          <div
                            aria-describedby="rbd-hidden-text-0-hidden-text-0"
                            data-rbd-drag-handle-context-id="0"
                            data-rbd-drag-handle-draggable-id="1"
                            data-rbd-draggable-context-id="0"
                            data-rbd-draggable-id="1"
                            draggable="false"
                            role="button"
                            tabindex="0"
                          >
                            Supported Languages
                          </div>
                          <svg
                            aria-hidden="true"
                            class="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
                            focusable="false"
                            viewBox="0 0 24 24"
                          >
                            <path
                              d="M4 12l1.41 1.41L11 7.83V20h2V7.83l5.58 5.59L20 12l-8-8-8 8z"
                            />
                          </svg>
                        </span>
                      </th>
                      <th
                        class="MuiTableCell-root MuiTableCell-head ForwardRef(MTableHeaderRef)-header-20 BackstageTableHeader-header-19 MuiTableCell-alignLeft"
                        scope="col"
                        style="text-transform: uppercase; box-sizing: border-box;"
                      >
                        <span
                          aria-disabled="false"
                          class="MuiButtonBase-root MuiTableSortLabel-root"
                          role="button"
                          tabindex="0"
                        >
                          <div
                            aria-describedby="rbd-hidden-text-0-hidden-text-0"
                            data-rbd-drag-handle-context-id="0"
                            data-rbd-drag-handle-draggable-id="2"
                            data-rbd-draggable-context-id="0"
                            data-rbd-draggable-id="2"
                            draggable="false"
                            role="button"
                            tabindex="0"
                          >
                            Builder
                          </div>
                          <svg
                            aria-hidden="true"
                            class="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
                            focusable="false"
                            viewBox="0 0 24 24"
                          >
                            <path
                              d="M4 12l1.41 1.41L11 7.83V20h2V7.83l5.58 5.59L20 12l-8-8-8 8z"
                            />
                          </svg>
                        </span>
                      </th>
                    </tr>
                  </thead>
                  <tbody
                    class="MuiTableBody-root"
                  >
                    <tr
                      class="MuiTableRow-root"
                      index="0"
                      level="0"
                    >
                      <td
                        class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignLeft"
                        icons="[object Object]"
                        style="box-sizing: border-box; font-family: inherit; font-weight: inherit;"
                        value="item.buildpack"
                      >
                        item.buildpack
                      </td>
                      <td
                        class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignLeft"
                        icons="[object Object]"
                        style="box-sizing: border-box; font-family: inherit; font-weight: inherit;"
                        value="item.languages"
                      >
                        item.languages
                      </td>
                      <td
                        class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignLeft"
                        icons="[object Object]"
                        style="box-sizing: border-box; font-family: inherit; font-weight: inherit;"
                        value="item.builder"
                      >
                        item.builder
                      </td>
                    </tr>
                    <tr
                      class="MuiTableRow-root"
                      index="1"
                      level="0"
                    >
                      <td
                        class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignLeft"
                        icons="[object Object]"
                        style="box-sizing: border-box; font-family: inherit; font-weight: inherit;"
                        value="item.buildpack2"
                      >
                        item.buildpack2
                      </td>
                      <td
                        class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignLeft"
                        icons="[object Object]"
                        style="box-sizing: border-box; font-family: inherit; font-weight: inherit;"
                        value="item.languages2"
                      >
                        item.languages2
                      </td>
                      <td
                        class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignLeft"
                        icons="[object Object]"
                        style="box-sizing: border-box; font-family: inherit; font-weight: inherit;"
                        value="item.builder2"
                      >
                        item.builder2
                      </td>
                    </tr>
                  </tbody>
                </table>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    aria-atomic="true"
    aria-live="assertive"
    id="rbd-announcement-0"
    style="position: absolute; width: 1px; height: 1px; margin: -1px; border: 0px; padding: 0px; overflow: hidden; clip-path: inset(100%);"
  />
  <div
    id="rbd-hidden-text-0-hidden-text-0"
    style="display: none;"
  >
    
  Press space bar to start a drag.
  When dragging you can use the arrow keys to move the item around and escape to cancel.
  Some screen readers may require you to be in focus mode or to use your pass through key

  </div>
</body>
`;
